File: [1043][Medium][Partition_Array_for_Maximum_Sum]/Partition_Array_for_Maximum_Sum.py
# leetcode time     cost : 640 ms
# leetcode memory   cost : 13.7 MB 
# solution 1, DP
class Solution:
    def maxSumAfterPartitioning(self, A, K) :
            n=len(A)
            dp=[0]*(n+1)
            for i in range(1,n+1) :
                j=i-1
                domain_max = float("-inf")
                while i-j <= K and j >= 0 :
                    #包含A[j] 的 last domain 所有可能，取最大的
                    domain_max = max(domain_max, A[j])
                    dp[i] = max(dp[i], dp[j] + domain_max *(i-j)) 
                    j-=1
            return dp[n]
    
def main():
    A, K = [1,15,7,9,2,5,10], 3         # expect is 84
    obj = Solution()
    result = obj.maxSumAfterPartitioning(A, K)
    print("return result is :",result)
    
if __name__ =='__main__':
    main() 

File: [1043][Medium][Partition_Array_for_Maximum_Sum]/Partition_Array_for_Maximum_Sum.java
// leetcode time     cost : 8 ms
// leetcode memory   cost : 39.6 MB 
// Time  Complexity: O(A.length*K)
// Space Complexity: O(A.length)
// solution 1, DP
class Solution {
    public int maxSumAfterPartitioning(int[] A, int K) {
        int len = A.length;
        int[] dp = new int[len];
        for (int i = 0; i < len; i++) {
            /* 分别计算最后一段区间长度 j ∈[1, K]时的解，并更新位置i时的最优解 */
            int domainMax = A[i];
            for (int j = 1; j <= K && i - j + 1 >= 0; j++) {
                domainMax = Math.max(domainMax, A[i-j+1]);
                if (i - j >= 0) {
                    dp[i] = Math.max(dp[i], dp[i-j] + j*domainMax);
                } else {
                    dp[i] = Math.max(dp[i], j*domainMax);
                }
            }
        }
        return dp[len-1];
    }
}